kv/txn_coord_sender.go

```diff
@@ -54,22 +55,27 @@ import (
 // periodically on their own.
 type txnMetadata struct {
 	// txn is the transaction struct from the initial AddRequest call.
-	txn proto.Transaction
+	txn *proto.Transaction
 
 	// keys stores key ranges affected by this transaction through this
 	// coordinator. By keeping this record, the coordinator will be able
 	// to update the write intent when the transaction is committed.
 	keys *cache.IntervalCache
 
-	// lastUpdateTS is the latest time when the client sent transaction
-	// operations to this coordinator.
-	lastUpdateTS proto.Timestamp
+	// lastUpdateNanos is the latest wall time in nanos the client sent
+	// transaction operations to this coordinator. Accessed and updated
+	// atomically.
+	lastUpdateNanos int64
 
 	// timeoutDuration is the time after which the transaction should be
 	// considered abandoned by the client. That is, when
 	// current_timestamp > lastUpdateTS + timeoutDuration If this value
 	// is set to 0, a default timeout will be used.
 	timeoutDuration time.Duration
+
+	// closeCh is closed when the transaction has completed to exit
```

Comment:
something off about this sentence. "closeCh is closed when the transaction completes, which signals that heartbeats should stop"? maybe `closeCh` should be renamed to associated it with heartbeats, too.